### PR TITLE
Music box

### PR DESCRIPTION
**Pull request by SilverTactic**
So I created these music box commands where you can add songs by supplying a valid Youtube URL.

### CODE REVIEW

**Comment by panpawn** [comment]
We already use the TC command to add new info-box based commands like this, so we can remove that entire part of this imo @SilverTactic

**Comment by panpawn** [comment]
Oh wait, nevermind I misread this.  I'm testing this soon, this actually looks interesting...

**Comment by SilverTactic** [comment]
Alright, cool.
Anyways, I edited the shop command, so go ahead and check that out. Now all we need is a way for a user to be able to purchase an extra song whenever they like

**Comment by panpawn** [comment]
Could we make this use ``Gold.musicboxes`` so we can reference it outside this file in the future?

File: chat-plugins/musicbox.js [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+		});
+	});
+}
+
+exports.commands = {
+	mb: 'musicbox',
+	musicbox: function (target, room, user, connection, cmd) {
+		var cmds = {'help':1, 'add':1, 'remove':1, 'css':1, 'removeall':1, 'delete':1};
+		if (target && toId(target.split(' ')[0]) in cmds) {
+			if (typeof musicboxes[user.userid] !== 'object') return this.sendReply("You don't have a music box. Buy one from the shop!");
```
Comment (panpawn) [inline chat-plugins/musicbox.js]: Let's remember to use ``this.errorReply()`` for cases like this btw

File: chat-plugins/musicbox.js [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+						'<li>/' + cmd + ' removeall - Removes all songs from your music box.<br>' +
+						'<li>/' + cmd + ' css <em>CSS code</em> - Edits the CSS of your music box\'s buttons.<br>' +
+						'<li>/' + cmd + ' delete <em>User</em> - Deletes a user\'s music box. Requires ~.'
+					);
+					break;
+
+				case 'add':
+					if (!target || !target.trim()) return this.parse('/' + cmd + ' help');
+					validate(target)
+					.then(function (song) {
```
Comment (panpawn) [inline chat-plugins/musicbox.js]: Put this on the same line as the one above it

File: chat-plugins/musicbox.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+			var targetUser;
+			if (!toId(target)) targetUser = user.name;
+			else targetUser = Users.getExact(target) ? Users.getExact(target).name : target;
+			var box = musicboxes[toId(targetUser)];
+			if (!box) return this.sendReplyBox(targetUser + ' doesn\'t have a music box...');
+			if (!box.songs.length) return this.sendReplyBox(targetUser + '\'s music box is empty...');
+
+			var total = [];
+			box.songs.forEach(function (song) {
+				console.log(song.css);
```
Comment (panpawn) [inline chat-plugins/musicbox.js]: Why?

**Comment by panpawn** [comment]
@SilverTactic

**Comment by panpawn** [comment]
Um Siiilver...

We already have a music box command, this needs to remove that so please ``git rm chat-plugins/music-box.js``, because that's the existing plugin we have.  (You named this one ``chat-plugins/musicbox.js``)

**Comment by SilverTactic** [comment]
...right. lol. I'll do that

File: chat-plugins/musicbox.js (hunk 3 above)
Comment (SilverTactic) [inline chat-plugins/musicbox.js]: If you're not using the music box management sub-commands (I guess?), the command'll just show you your (or someone else's) music box

File: chat-plugins/musicbox.js (hunk 1 above)
Comment (SilverTactic) [inline chat-plugins/musicbox.js]: Okie

File: chat-plugins/musicbox.js (hunk 3 above)
Comment (SilverTactic) [inline chat-plugins/musicbox.js]: If you're talking about the <code>console.log</code> thing, then yeah, I need to remove that lol. I used it to test something

Comment (panpawn) [inline chat-plugins/musicbox.js]: yea that's what I was talking about

File: chat-plugins/musicbox.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+					fs.writeFileSync(FILE, JSON.stringify(musicboxes, null, 1));
+					return this.sendReply('|html|The song "<b>' + match.title + '</b>" has been successfully removed from your music box.');
+					break;
+					
+				case 'css':
+					var box = musicboxes[user.userid];
+					if (!target || !target.trim()) {
+						if (!this.canBroadcast()) return;
+						if (toId(box.css)) return this.sendReplyBox('Your music box css: <code>' + box.css + '</code>');
+						return this.sendReplyBox('You haven\'t set button css for your music box yet.');
```
Comment (panpawn) [inline chat-plugins/musicbox.js]: Use ``this.errorReply()`` here

File: chat-plugins/musicbox.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+				case 'delete':
+					if (!this.can('hotpatch')) return false;
+					var targetUser = Users.getExact(target) ? Users.getExact(target).name : target;
+					var box = musicboxes[toId(targetUser)];
+					if (!box) return this.sendReply(targetUser + ' doesn\'t have a music box...');
+
+					delete box;
+					fs.writeFileSync(FILE, JSON.stringify(musicboxes, null, 1));
+					return this.sendReply('You have successfully deleted ' + targetUser + '\'s music box.');
```
Comment (panpawn) [inline chat-plugins/musicbox.js]: Missing a ``break;`` statement here it seems?

File: chat-plugins/musicbox.js (hunk 4 above)
Comment (SilverTactic) [inline chat-plugins/musicbox.js]: I used <code>this.sendReplyBox()</code> because that part's broadcastable

File: chat-plugins/musicbox.js (hunk 5 above)
Comment (SilverTactic) [inline chat-plugins/musicbox.js]: It isn't really necessary because it's the last switch statement, but okay

**Comment by panpawn** [comment]
Can we give all of these commands to leaders as well?  Leaders should have full access to removing these and whatnot...

``if (!this.can('pban')) return false;``

**Comment by SilverTactic** [comment]
Alright.
Could you gimme a list of everything else that needs to be edited too? I'm too lazy to keep amending the commit xD

**Comment by panpawn** [comment]
What has to be done...
- Give leaders and up all of the higher auth permissions on this
- Remove old music box command file to be in replaced of this one
- Give me time to test after those changed have been made

**Comment by SilverTactic** [comment]
Okie done~
I still haven't decided whether or not I should separate the commands...

**Pull request merged**